lucene/expressions/src/test/org/apache/lucene/expressions/TestExpressionValidation.java

```diff
@@ -110,4 +110,15 @@ public void testCoRecursion4() throws Exception {
     });
     assertTrue(expected.getMessage().contains("Cycle detected"));
   }
+
+  public void testCoRecursion42() throws Exception {
```

Comment:
I provided this test but I don't think we should add it as it relies on iteration order and might get defeated on some JVMs or future versions of Java. I'd suggest to not add this test, or to change the map in SimpleBindings from a HashMap to a TreeMap to be able to better test such cases (in which case we'd have to swap `cycle0`/`cycle2`), this could be done in a follow-up too.